File: apps/pass-at-k-study/benchmark/tasks/legate/l3_partitioned_exchange_validate.py
# Task: l3_partitioned_exchange_validate — matches L3PartitionedExchangeValidate.
import sys

_s = 11


def randf01() -> float:
    global _s
    _s = (_s * 1103515245 + 12345) & 0x7FFFFFFF
    return float(_s & 0xFFFFFF) / float(0x1000000)


def run(n: int) -> bool:
    a = [[randf01() for _ in range(n)] for _ in range(n)]
    r = [[a[i][j] + a[j][i] for j in range(n)] for i in range(n)]
    for t in range(5):
        i = (t * 17) % n
        j = (t * 31) % n
        if abs(r[i][j] - (a[i][j] + a[j][i])) > 1e-4:
            return False
    return True


def main() -> int:
    global _s
    _s = 11
    if not run(128):
        print("FAIL")
        return 1
    _s = 12
    if not run(512):
        print("FAIL")
        return 1
    print("PASS")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: apps/pass-at-k-study/benchmark/tasks/iris/l3_partitioned_exchange_validate.cpp
// Task: l3_partitioned_exchange_validate — matches L3PartitionedExchangeValidate.
#include <cmath>
#include <cstdint>
#include <iostream>
#include <vector>

static std::uint32_t g_seed = 11u;
static float randf01() {
  g_seed = g_seed * 1103515245u + 12345u;
  return float(g_seed & 0xffffffu) / float(0x1000000u);
}

static bool run(int N) {
  std::vector<float> A(size_t(N) * N), R(size_t(N) * N);
  for (int i = 0; i < N; ++i)
    for (int j = 0; j < N; ++j) A[size_t(i) * N + j] = randf01();
  for (int i = 0; i < N; ++i)
    for (int j = 0; j < N; ++j)
      R[size_t(i) * N + j] = A[size_t(i) * N + j] + A[size_t(j) * N + i];
  for (int t = 0; t < 5; ++t) {
    int i = (t * 17) % N;
    int j = (t * 31) % N;
    float got = R[size_t(i) * N + j];
    float exp = A[size_t(i) * N + j] + A[size_t(j) * N + i];
    if (std::fabs(got - exp) > 1e-4f) return false;
  }
  return true;
}

int main() {
  g_seed = 11u;
  if (!run(128)) {
    std::cout << "FAIL\n";
    return 1;
  }
  g_seed = 12u;
  if (!run(512)) {
    std::cout << "FAIL\n";
    return 1;
  }
  std::cout << "PASS\n";
  return 0;
}
